packages/babel-core/src/config/files/module-types.js

```diff
@@ -0,0 +1,57 @@
+import { isAsync, waitFor } from "../../gensync-utils/async";
+import type { Handler } from "gensync";
+
+let import_;
+try {
+  // Node < 13.3 doesn't support import() syntax.
+  import_ = require("./import").default;
+} catch {}
+
+export default function* loadCjsOrMjsDefault(
+  filepath: string,
+  asyncError: string,
+): Handler<mixed> {
+  switch (guessJSModuleType(filepath)) {
+    case "cjs":
+      return loadCjsDefault(filepath);
+    case "unknown":
+      try {
+        return loadCjsDefault(filepath);
+      } catch (e) {
+        if (e.code !== "ERR_REQUIRE_ESM") throw e;
+      }
+    case "mjs":
+      if (yield* isAsync()) {
+        return yield* waitFor(loadMjsDefault(filepath));
+      }
+      throw new Error(asyncError);
+  }
+}
+
+function guessJSModuleType(path: string): "cjs" | "mjs" | "unknown" {
+  switch (path.slice(-4)) {
+    case ".cjs":
+      return "cjs";
+    case ".mjs":
+      return "mjs";
+    default:
+      return "unknown";
+  }
+}
+
+function loadCjsDefault(filepath: string) {
+  const module = (require(filepath): mixed);
+  return module?.__esModule ? module.default || undefined : module;
```

Comment:
Why `undefined` is offered as fallback here?

Reply:
I'm not sure; I only moved this code here from https://github.com/babel/babel/blob/5b907e9bb7e356fe869aaa9e3f8aed8c4991f1b5/packages/babel-core/src/config/files/configuration.js#L181-L185

Reply:
Maybe in case someone exports `false` as their default config? :thinking:

Reply:
Alright, we should investigate this in later PRs, or just remove it in Babel 8 because it looks suspicious.